feat: handle optional alg parameter for keys

froissant opened the pull request:
As per the JSON Web Key specification (RFC 7517), the [alg parameter is optional](https://tools.ietf.org/html/rfc7517#section-4.4).

Some providers, for example Azure AD, don't provide the algorithm, so retrieveJwk always return errors.JWK_NOT_FOUND.

I've added a simple check for the property in the keys.find() filter.

mcollina reviewed (approved):
lgtm

simoneb reviewed (changes requested):
can you please add a test to cover this case?

froissant commented:
I updated the existing test cases with an additional key that lacks the alg property to cover this case, what other kind of test do you have in mind? Thanks 🙂

simoneb commented:
> I updated the existing test cases with an additional key that lacks the alg property to cover this case, what other kind of test do you have in mind? Thanks 🙂

The change you've done there does nothing, because you're using a key with an alg. You should add test(s) to cover the case where the key you're retrieving doesn't have an alg. See the existing automated tests

froissant commented:
> > I updated the existing test cases with an additional key that lacks the alg property to cover this case, what other kind of test do you have in mind? Thanks 🙂
> 
> The change you've done there does nothing, because you're using a key with an alg. You should add test(s) to cover the case where the key you're retrieving doesn't have an alg. See the existing automated tests

OK I see what I did wrong when trying to make the code a bit "cleaner", I updated it to avoid side-effects. If I change `const jwk = jwks.keys[1]` to `const jwk = jwks.keys[2]`, then it will use KEY_2 and it doesn't have the `alg` property.
I would like to be able to run the tests with all key configurations but I'm not sure how to do it, I'm not very familiar with tap.

simoneb commented:
@froissant your test code is doing nothing. you're operating on keys[1], which has always an algorithm, so I'm not sure what you're trying to do. see https://github.com/nearform/get-jwks/pull/45/files#diff-366f6c7f36704ace098bd7d73c82ffe77a5c8cd61f3bae882dd5a06a6772c66eR40.

the production code instead looks like it's doing something and it looks correct, but you need to _write unit tests_ to cover that scenario

froissant commented:
> @froissant your test code is doing nothing. you're operating on keys[1], which has always an algorithm, so I'm not sure what you're trying to do. see https://github.com/nearform/get-jwks/pull/45/files#diff-366f6c7f36704ace098bd7d73c82ffe77a5c8cd61f3bae882dd5a06a6772c66eR40.
> 
> the production code instead looks like it's doing something and it looks correct, but you need to _write unit tests_ to cover that scenario

I've added 2 unit tests based on the existing ones, is that OK for you?

I didn't introduce the keys[1] hardcoding, and I think it would make more sense if we ran all the tests supplying the different keys, but as I mentioned previously I'm not sure what the best implementation for this would be.

Let me know if you have any comments.

simoneb commented on test/constants.js:
```diff
@@ -26,12 +26,24 @@ const jwks = {
 const privateKey = readFileSync(path.join(__dirname, 'private.pem'), 'utf8')
 const jwk = jwks.keys[1]
+let defaultSigningAlgorithm;
+if (jwk.alg === undefined && jwk.kty === 'RSA') {
+  defaultSigningAlgorithm = 'RS256';
+}
 const token = jwt.sign({ name: 'Jane Doe' }, privateKey, {
-  algorithm: jwk.alg,
+  algorithm: jwk.alg || defaultSigningAlgorithm,
```
as far as I understand this code is doing nothing. am I missing something? can you try to remove it and see what happens?

froissant replied:
If you replace `keys[1]` by `keys[2]`, then jwt.sign will fail because there is no algorithm to sign with (_Error: "algorithm" must be a valid string enum value_) ; I had the idea that we could perform the test with differents keys, but since I was mistaken I've removed it 🙂

simoneb commented on test/constants.js:
```diff
@@ -40,6 +52,8 @@ const oidcConfig = {
   jwks_uri: 'https://localhost/.well-known/certs',
 }
 
+console.log(jwks)
```
remove

mcollina reviewed (approved):
lgtm

simoneb commented:
Code wise this looks good now, thanks. Can you please update the readme to clarify that this parameter is optional?

froissant commented:
Let me know if you need anything else :)

simoneb reviewed (approved):
LGTM. Next time hit the re-request review button so we're informed when you're addressing review feedback :)

Thanks for this!

froissant commented:
> LGTM. Next time hit the re-request review button so we're informed when you're addressing review feedback :)
> 
> Thanks for this!

Noted, thanks!
